src/Microsoft.AspNetCore.Server.IISIntegration/Server/HttpProtocol.cs

```diff
@@ -530,7 +556,21 @@ public void StartReadingRequestBody()
 
                     try
                     {
-                        int read = await ReadAsync(wb.Buffer.Length);
+                        int read = 0;
+                        if (_wasUpgraded)
+                        {
+                            await ReadAsync(wb.Buffer.Length);
+                        }
+                        else
+                        {
+                            currentOperation = currentOperation.ContinueWith(async (t) =>
```

Comment:
We need to remove the closure allocations here. ContinueWith takes a state object. You may need to store the amount read and the length in a field. This is what the pattern looks like:

```C#
private static Func<Task, object, Task> _doReadAsync = DoReadAsync;
private int _read;
private int _length;

private static Task DoReadAsync(Task previous, object state)
{
    return ((HttpProtocol)state).DoReadAsync();
}

private async Task DoReadAsync()
{
    _currentAsyncOperation = CurrentOperationType.Read;
    _read = await ReadAsync(wb.Buffer.Length);
}


currentOperation = currentOperation.ContinueWith(_doReadAsync, this).Unwrap();

await currentOperation;
```

We might also want to come up with something less allocatey once we get it all working.

Reply:
#451